framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java

```diff
@@ -267,97 +268,176 @@ protected boolean areEqualInHierarchy(
-            outsideLowerBound = ((AnnotatedTypeVariable) outside).getLowerBound();
-        } else {
-            throw new BugInCF(
-                    "Expected a wildcard or captured type variable, but found " + outside);
+            typeargVisitHistory.add(inside, outside, currentTop, true);
+            boolean result = isContainedByBoundType(inside, new BoundType(outside), canBeCovariant);
+            typeargVisitHistory.add(inside, outside, currentTop, result);
+            return result;
         }
-        Boolean previousResult = typeargVisitHistory.result(inside, outside, currentTop);
-        if (previousResult != null) {
-            return previousResult;
+        if (canBeCovariant) {
+            return isSubtype(inside, outside, currentTop);
         }
-
-        typeargVisitHistory.add(inside, outside, currentTop, true);
-        boolean result =
-                isContainedWildcard(
-                        inside, outside, outsideUpperBound, outsideLowerBound, canBeCovariant);
-        typeargVisitHistory.add(inside, outside, currentTop, result);
-        return result;
+        return areEqualInHierarchy(inside, outside);
     }
 
-    private boolean isContainedWildcard(
-            AnnotatedTypeMirror inside,
-            AnnotatedTypeMirror outside,
-            AnnotatedTypeMirror outsideUpperBound,
-            AnnotatedTypeMirror outsideLowerBound,
-            boolean canBeCovariant) {
-
-        if (inside.equals(outside)) {
-            // If they are equal, outside always contains inside.
-            return true;
-        }
-
-        if (inside.getKind() == TypeKind.WILDCARD) {
-            outsideUpperBound =
-                    checker.getTypeFactory()
-                            .widenToUpperBound(outsideUpperBound, (AnnotatedWildcardType) inside);
-        }
-        while (outsideUpperBound.getKind() == TypeKind.WILDCARD) {
-            if (ignoreUninferredTypeArgument(outsideUpperBound)) {
-                return true;
+    /**
+     * Returns true if {@code outside} contains {@code inside}, that is, if the set of types denoted
+     * by {@code outside} is a superset of or equal to the set of types denoted by {@code inside}.
+     *
+     * <p>A declared type is considered a supertype of another declared type only if all of the type
+     * arguments of the declared type "contain" the corresponding type arguments of the subtype.
+     * Containment is described in <a
+     * href="https://docs.oracle.com/javase/specs/jls/se11/html/jls-4.html#jls-4.5.1">JLS section
+     * 4.5.1 "Type Arguments of Parameterized Types"</a>.
+     *
+     * @param inside a possibly-contained type
+     * @param outside a possibly-containing type
+     * @param canBeCovariant whether or not type arguments are allowed to be covariant
+     * @return true if inside is contained by outside, or if canBeCovariant == true and {@code
+     *     inside <: outside}
+     */
+    protected boolean isContainedByBoundType(
```

Comment:
Does this need to be protected, or can it be public?

Reply:
I could make it public, but there isn't much point.  Clients of DefaultTypeHierarchy access it via the interface TypeHierarchy whose only method is isSubtype.